qubes/api/admin.py

```diff
@@ -119,11 +140,10 @@ def vm_list(self):
         else:
             domains = self.fire_event_for_filter([self.dest])
 
-        return ''.join('{} class={} state={}\n'.format(
-                vm.name,
-                vm.__class__.__name__,
-                vm.get_power_state())
-            for vm in sorted(domains))
+        strs = []
+        for vm in sorted(domains):
+            self._vm_line_strs(strs, vm)
+        return ''.join(strs)
```

Comment:
Is `str.append` really faster than `str.format`?

Reply:
The idea is to avoid doing any kind of string processing until the final join call, which should be the fastest since it should just allocate a single string in join and copy them all in, avoiding any intermediate allocations, partial concatenations, format string parsing, etc.

It's slightly uglier, but not that much, it looks like writing to an output stream (which would be even faster, but not really worth the effort).